### PR TITLE
[Fix 0022298] general PostgreSQL compatiblity

### PR DESCRIPTION
**Pull request by lbruening**
I've made several changes to enable postgresql-compatiblity again, so setup will work without errors... Feel free to check

### CODE REVIEW

**Comment by lbruening** [comment]
Is this really an improvement rather than a bugfix? Postgres-Support is totally broken in current 5.3-branch and in trunk. Field names are quoted hard-coded in several scripts.

**Comment by Amstutz** [comment]
@lbruening Thank you for reporting this. If you think that this classifies as but, you need to create an according ticket on mantis an link this here (see: https://github.com/ILIAS-eLearning/ILIAS/blob/trunk/docs/documentation/contributing.md#rules-for-contributors). Otherwise your request will be labeled as improvement.

**Comment by lbruening** [comment]
i opened a ticket on matis for this... (0022298)

File: setup/sql/dbupdate_04.php
Comment (chfsx) [inline setup/sql/dbupdate_04.php:18543]: I think this should be `integer` not `text` and value then should be `1`instead of `"1"`.

File: Services/Database/classes/PDO/class.ilDBPdo.php
Comment (chfsx) [inline Services/Database/classes/PDO/class.ilDBPdo.php:614]: this needs a loop with `$this->quoteIdentifier( ... )` to avoid problems with reserved words.

File: Modules/Scorm2004/data/seq_editor/seqtemplates.sql.php
Comment (chfsx) [inline Modules/Scorm2004/data/seq_editor/seqtemplates.sql.php]: @Uwe-Kohnle what is this file about? could you please move this to proper db-update-steps if really needed (I did not found usages of this file in trunk).

File: Modules/TestQuestionPool/classes/class.assClozeGapCombination.php
Comment (chfsx) [inline Modules/TestQuestionPool/classes/class.assClozeGapCombination.php:188]: @bheyser ok for you? quoting the table-name isn't really needed here.

File: Services/ActiveRecord/Connector/class.arConnectorDB.php
Comment (chfsx) [inline Services/ActiveRecord/Connector/class.arConnectorDB.php:73]: Thanks!

File: setup/sql/dbupdate_04.php
Comment (chfsx) [inline setup/sql/dbupdate_04.php]: @mjansenDatabay is this ok for you? LGTM

File: setup/sql/dbupdate_04.php
Comment (chfsx) [inline setup/sql/dbupdate_04.php]: @smeyer-ilias or @alex40724 , this LGTM, how about you?

**Review by chfsx — CHANGES_REQUESTED** [review]
Thank you very much for your work which looks good to me in general, there are some small changes required and I asked all involved maintainers to check their parts as well.

File: Services/Database/classes/PDO/class.ilDBPdo.php
Comment (lbruening) [reply Services/Database/classes/PDO/class.ilDBPdo.php:614]: fields are quouted in foreach-loop at line 601

File: setup/sql/dbupdate_04.php
Comment (lbruening) [reply setup/sql/dbupdate_04.php:18543]: fp.value (il_dcl_fieldprop) is "varchar" and postgresql raised an error wenn try to compare to an integer

File: setup/sql/dbupdate_04.php
Comment (alex40724) [reply setup/sql/dbupdate_04.php]: The lines do no harm. But I wonder why they are necessary. Does renameTable only work without primary keys on postgres? This would be different from MySQL, so the DB classes should take care of this imo.

Comment (lbruening) [reply setup/sql/dbupdate_04.php]: Primary Keys in PostgreSQL have an unique name rather than in MySQL, so renaming a table doesn't  rename the constraint.

In step 5168 of dbupdate the primary key of auth_ext_attr_mapping will be dropped, but the constraint-name  is still called saml_attribute_mapping_(pk_idx).

Comment (lbruening) [reply setup/sql/dbupdate_04.php]: ok, maybe i should revert the recreation of primary keys and rename them automatically when renaming the table? what about other constrains and indexes? should they renamed also?

Comment (chfsx) [reply setup/sql/dbupdate_04.php]: If you find the time to investigate how this can be done the best way, I‘d really appreciate since I currently have no running Postgres-based installation running. Thank you very much!

**Review by lbruening — COMMENTED** [review]
I reverted the recreation of primary keys in dbupdate_04. Now renaming a table will also rename all indexes and constraints.

File: Services/MyStaff/classes/ListUsers/class.ilMStListUsers.php
Comment (chfsx) [inline Services/MyStaff/classes/ListUsers/class.ilMStListUsers.php:58]: @mstuder OK for you?

**Comment by chfsx** [comment]
LGTM in general, waiting for feedback from some other developers which have steaks in here...

**Review by chfsx — APPROVED** [review]

**Comment by chfsx** [comment]
I got feedback from @Uwe-Kohnle via email. currently no feedback from @bheyser. Since the changes in his code are very small, I merge this anyway since I really appreciate the changes and the proposal you made! Thank you very much, this is a very important PR! A colleague of mine testet the installation with Postgres with your changes and it works very fine. We will also propose a installation guide for postgres, maybe you @lbruening could have a look at this as well? it's this one: https://github.com/studer-raimann/ILIAS/blob/proposal/5-4/installation-postgres/docs/configuration/postgresql.md

**Comment by chfsx** [comment]
oh I just see there are some new changes in `ilTree` and `ilMaterializedPathTree`. These are very central classes and not in my scope of maintenance. Sorry @lbruening we have to ask first if these changes are ok, too. you could also revert this and open a new PR for just those changes.

File: Services/Tree/classes/class.ilMaterializedPathTree.php
Comment (chfsx) [inline Services/Tree/classes/class.ilMaterializedPathTree.php:229]: @smeyer-ilias could you please review these changes in the Tree-Classes? Thank you very much

File: Services/Tree/classes/class.ilMaterializedPathTree.php
Comment (chfsx) [inline Services/Tree/classes/class.ilMaterializedPathTree.php]: @smeyer-ilias could you please review these changes in the Tree-Classes? Thank you very much

File: Services/Tree/classes/class.ilTree.php
Comment (chfsx) [inline Services/Tree/classes/class.ilTree.php:2849]: @smeyer-ilias this seems OK from my point of view

**Comment by lbruening** [comment]
@chfsx i think those commits should stay in this pr, because they are essential to run/install Ilias under postgres. I had a look at your install-instructions. LGTM
In my research i found some other sources where changes must be made in order to run Ilias properly under postgres. I will commit these changes later in a new pr.

File: Modules/TestQuestionPool/classes/class.assClozeGapCombination.php
Comment (bheyser) [reply Modules/TestQuestionPool/classes/class.assClozeGapCombination.php:188]: this is okay for me :+1:

**Review by bheyser — APPROVED** [review]
approved my part of the PR

**Comment by matthiaskunkel** [comment]
Jour Fixe, 05 FEB 2018: We wait for the review of ilTree by Stefan Meyer. If accepted, this PR can be merged to trunk, 5.3 and 5.2.

**Review by smeyer-ilias — APPROVED** [review]
@lbruening thanks for this! The tree changes look good to me.

File: Services/Tree/classes/class.ilMaterializedPathTree.php
Comment (smeyer-ilias) [reply Services/Tree/classes/class.ilMaterializedPathTree.php]: Have to revert my approval.
The part "cast(child as varchar)" is not standard SQL.

**Review by smeyer-ilias — CHANGES_REQUESTED** [review]
see comment for ilMaterializedPathTree.php

**Review by chfsx — CHANGES_REQUESTED** [review]
@lbruening please have a look at the second comment from @smeyer-ilias

File: Services/Tree/classes/class.ilMaterializedPathTree.php
Comment (lbruening) [reply Services/Tree/classes/class.ilMaterializedPathTree.php]: CAST(FIELD AS TYPE) is part of SQL92...
The Problem is, that postgres needs explicit casts (concat with different types)

Comment (chfsx) [reply Services/Tree/classes/class.ilMaterializedPathTree.php]: At least this will fail with Oracle I think, there it should be a `VARCHAR2`. This Query is tricky... I'm currently thinking of abstracting `CAST` but currently there is no Oracle-based installation I could test this.

Comment (lbruening) [reply Services/Tree/classes/class.ilMaterializedPathTree.php]: ok, now whe have two options:
- find a sql-statement that will work on all databases
- abstract the method "`cast`"

for the first option this should work:
```SQL
COALESCE(TRIM(CAST(child AS CHAR(2000))), '')
```

for the second option: 
i can implement a new method in ilDBInterface similar to groupConcat in PR #866

what are your preferences?

Comment (smeyer-ilias) [reply Services/Tree/classes/class.ilMaterializedPathTree.php]: @lbruening Yes, your code is compliant to SQL92, but the data types are handled differently (MySQL, Oracle). 

MySQL: 
cast(something as varchar) is not supported
cast(something as char) or cast(something as char(4000)) is supported

Oracle [1]:
select cast(banner as char(4000)) from v$version; => size to big for datatype
select cast(banner as char(1000)) from v$version; => ok

[1] https://www.tutorialspoint.com/oracle_terminal_online.php

Comment (lbruening) [reply Services/Tree/classes/class.ilMaterializedPathTree.php]: ...and in Postgres, a cast(something as char(4000)) will return a string of size 4000 (padded with spaces.... damned!)

@chfsx 
ok, i think, i will create an abstraction of the cast-method in ilDBInterface... is this ok for you?

Comment (chfsx) [reply Services/Tree/classes/class.ilMaterializedPathTree.php]: @lbruening absolutely, and thank you very much for all your time you invest in this!

**Comment by lbruening** [comment]
I made a new commit. Please review.

**Comment by matthiaskunkel** [comment]
Jour Fixe, 12 MAR 2018: Stefan will have a final look at the latest changes. PHP 7 specific code should be removed.

File: Services/Database/classes/MDB2/class.ilDB.php
Comment (chfsx) [inline Services/Database/classes/MDB2/class.ilDB.php]: @lbruening just saw you use PHP7 syntax here, could you please revert this to PHP5.x Synthat without return type declaration. thanks in advance!

File: Services/Database/interfaces/interface.ilDBInterface.php
Comment (chfsx) [inline Services/Database/interfaces/interface.ilDBInterface.php]: `$a_dest_type` will be a string since we pass the Abstract types

File: Services/Database/classes/QueryUtils/class.ilPostgresQueryUtils.php
Comment (chfsx) [inline Services/Database/classes/QueryUtils/class.ilPostgresQueryUtils.php:296]: I'd like to only pass the $a_dest_type as a string, not string or array.

**Review by chfsx — CHANGES_REQUESTED** [review]
@lbruening I really apologize for all the time you has to wait. I'll merge this this week after the following steps:

- I just saw there is PHP 7 syntax here, could you please remove this?
- @smeyer-ilias will look at the cast()-usage in his code, the abstraction LGT with one small change: please just use strings for ` $a_dest_type`

**Review by chfsx — APPROVED** [review]
Thanks!!

**Comment by chfsx** [comment]
Thank you very much @lbruening and welcome to the community ;)

**Pull request merged**
